### PR TITLE
HPCC-14442 Initial work on global/slave memory managers

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith please review.  I believe this is fully functional.

**Comment by ghalliday** [comment]
One oddity is that I have used std::vector instead of StructArrayOf.  It was primarily an experiment, and I can easily replace it with the jarray version.

**Comment by richardkchapman** [comment]
@jakesmith This seems to be stalled?

File: roxie/roxiemem/roxiemem.cpp [hunk 1]
```diff
@@ -2922,14 +2941,93 @@ class BufferedRowCallbackManager
...
+        typedef std::pair<unsigned, IBufferedRowCallback *> CallbackPair;
+        typedef std::vector<CallbackPair> CallbackPairArray;
+
+    public:
+        CallbackItem(unsigned _cost, unsigned _activityId) : cost(_cost), activityId(_activityId), nextCallback(0)
+        {
+        }
+        ~CallbackItem()
+        {
+        }
```
Comment (jakesmith) [inline roxie/roxiemem/roxiemem.cpp]: style:(?) - not needed.

File: thorlcr/activities/rollup/throllupslave.cpp [hunk 2]
```diff
@@ -568,6 +572,9 @@ class CRollupGroupSlaveActivity : public CSlaveActivity, public CThorDataLink
 
     CRollupGroupSlaveActivity(CGraphElementBase *_container) : CSlaveActivity(_container), CThorDataLink(this)
     {
+        eoi = false;
+        input = NULL;
+        helper = NULL;
```
Comment (jakesmith) [inline thorlcr/activities/rollup/throllupslave.cpp]: arguably these and other initializers should go in different PR.

File: roxie/roxiemem/roxiemem.cpp [hunk 3]
```diff
@@ -2922,14 +2941,93 @@ class BufferedRowCallbackManager
...
+            throwUnexpectedX("Removing callback that has not been registered");
+        }
+        unsigned releaseRows(BufferedRowCallbackManager & manager, unsigned whichSlave, unsigned minSuccess, bool critical)
+        {
+            CallbackPairArray::const_iterator begin = callbacks.begin();
+            CallbackPairArray::const_iterator end = callbacks.end();
+            CallbackPairArray::const_iterator iter = begin;
+            //MORE: Should nextCallback be tracked per slave?  There are arguments for and against.
+            if (nextCallback < callbacks.size())
+                iter += nextCallback;
```
Comment (jakesmith) [inline roxie/roxiemem/roxiemem.cpp]: Is this safe? Could it be out of bounds if num callbacks has shrunk between calls?

**Comment by jakesmith** [comment]
Minor comments, looks good to me.
@ghalliday

File: roxie/roxiemem/roxiemem.cpp (hunk 3 above)
Comment (ghalliday) [inline roxie/roxiemem/roxiemem.cpp]: I don't think so - that was part of the intention of the test.  If callbacks has shrunk then the test will fail, and iter will begin from the start of the list.

File: thorlcr/activities/rollup/throllupslave.cpp (hunk 2 above)
Comment (ghalliday) [inline thorlcr/activities/rollup/throllupslave.cpp]: yes, I agree.  I'll leave as is though, unless Richard wants them split.

**Comment by ghalliday** [comment]
@richardkchapman please check - released minor change in response to comments.

**Comment by richardkchapman** [comment]
Can you squash the commits please.

What if any conclusions did you draw from using std:vector etc ?

**Comment by richardkchapman** [comment]
@ghalliday

**Comment by ghalliday** [comment]
@richardkchapman I have switched back to using a jarray.  It is a separate commit, so you can see the differences.  I can squash before you merge.

**Comment by richardkchapman** [comment]
@ghalliday needs rebasing and squashing

**Comment by ghalliday** [comment]
@richardkchapman rebased and squashed - please review.
